### DISCUSSION TITLE
Is it possible do to ground state searching using a mixed state matrix ?

### DISCUSSION THREAD

**Discussion by ReiNicolas** [netket/netket#2222] [Q&A]
Hey, sorry for my naivity, I have been working on something and for that I needed to be able to expect_and_grad(Hamiltonian) using a variational mixed state, but wheneaver I do so I get an error message that the size for label i (n) does not match previus terms (2n), I presume this is some sort of non-compatibility between the samples and the hamiltonian.
 
 Here I have my code:

```python
L=4
g=nk.graph.Hypercube(length=L,n_dim=1,pbc=False)
hilbert = nk.hilbert.Spin(s=0.5, N=g.n_nodes)
double_hilbert=nk.hilbert.DoubledHilbert(hilbert)

H = nk.operator.Heisenberg(hilbert=hilbert, graph=g)

ma = Jastrow(N=double_hilbert.size, rngs=nnx.Rngs(0))
 
sampler=nk.sampler.MetropolisLocal(double_hilbert)

vs=nk.vqs.MCMixedState(sampler,model=ma)

#double_hilbert.all_states().reshape(2,2,-1)-double_hilbert.all_states().reshape(2,2,2)
vs.expect_and_grad(H)
```
 Now my questions is:
 1-What the hell am I doing wrong ? It works fine if I don't do expect_and_grad(H) but just expect(H).
 2-If the answer to 1 is no. Is this a bug or an I supposed to always do ground state searching using state-vector representations ?

**Comment by PhilipVinc** [comment]
`expect_and_grad` for mixed states and normal operators (as opposed to super-operators) is not implemented in netket. 

Note that what you probably need to do here is to derive the formula for the estimator you want to use, implement it as a function and register it as a dispatch rule as explained in the new error message.
I think I've never seen such an estimator used in any paper, but I might be forgetting something.
However do note that it is not entirely trivial, if I recall correctly, and there's multiple possible estimators you can choose from with different variance/bias tradeoffs. Identifying the good one will probably be a few days of work (and I'd suggest to do something like [Neural Projected Quantum Dynamics: a systematic study](https://quantum-journal.org/papers/q-2025-07-22-1803) to identify the best).

Nevertheless, there was a bug: you don't get a 'clean' error, instead you get something imperscrutable because netket is trying to use the standard purestate + operator gradient formula which makes no sense in this case.
I just pushed a change to print a better error message with a suggestion on what function to define to fix it.
To get it, just install from git with `uv add git+https://github.com/netket/netket`.

Your snippet now gives the following output:

```python

File ~/Nextcloud/Codes/Python/netket/netket/vqs/mc/mc_mixed_state/expect_grad.py:63, in expect_and_grad_operator_fallback(vstate, operator, chunk_size, mutable, *args, **kwargs)
     54 @expect_and_grad.dispatch
     55 def expect_and_grad_operator_fallback(  # noqa: F811
     56     vstate: MCMixedState,
   (...)     61     **kwargs,
     62 ):
---> 63     raise MCMixedStateExpectAndGradOnPhysicalOperatorError(vstate, operator)

MCMixedStateExpectAndGradOnPhysicalOperatorError:
            `expect_and_grad` is not implemented for `MCMixedState` with physical
            (non-super) operators.

            NetKet only provides built-in `MCMixedState.expect_and_grad` support for
            superoperators.

            To override this behaviour, define a more specific dispatch rule such as:

            @netket.vqs.expect_and_grad.dispatch
def expect_and_grad(
    vstate: netket.vqs.mc.mc_mixed_state.state.MCMixedState,
    operator: netket.operator._local_operator.jax.LocalOperatorJax,
    chunk_size: int | None,
    *args,
    mutable: CollectionFilter = False,
    **kwargs,
):
    ...


            In practice you will likely want to dispatch on `netket.operator._local_operator.jax.LocalOperatorJax`
            or, more likely, one of its supertypes rather than this exact concrete class.

            For background, references, and guidance on comparing multiple candidate
            estimators, see the online documentation linked below.


-------------------------------------------------------
For more detailed informations, visit the following link:
	 https://netket.readthedocs.io/en/latest/api/_generated/errors/netket.errors.MCMixedStateExpectAndGradOnPhysicalOperatorError.html
or the list of all common errors at
	 https://netket.readthedocs.io/en/latest/api/errors.html
-------------------------------------------------------

``

**Reply by ReiNicolas** [reply]
Thank you for your help and for the guidance.

**Comment by omni-front** [comment]
I had a similar issue when trying to use `expect_and_grad` with a mixed state matrix. In my case, the problem stemmed from the fact that `expect_and_grad` is not implemented for mixed states directly in NetKet. What happens is that when you run `expect_and_grad(H)`, the library expects the operator and the wave function to be compatible in form, which is why you're seeing that label size mismatch error.

When I faced this, I figured out that the main limitation is because `expect_and_grad` isn't designed to handle operators that aren't super-operators with mixed states. Instead, what worked for me was breaking down the problem into separate parts: first calculating the expectation value and then manually deriving the gradient separately. This involves working with the variational state to ensure it is consistent with the operator used.

If restructuring your calculations doesn't suit your needs, you might consider sticking to ground state searches using state-vector representations, which are more straightforward in the current NetKet implementation. Keep an eye on updates from the library, though, as features can evolve. But for now, manual intervention is required for mixed states with `expect_and_grad`.
